controller/cmd/service-mirror/cluster_watcher.go

```diff
@@ -497,6 +500,132 @@ func remapRemoteServicePorts(ports []corev1.ServicePort) []corev1.ServicePort {
 	return newPorts
 }
 
+func (rcsw *RemoteClusterServiceWatcher) updateGatewayProxyService(spec *GatewaySpec) error {
+	localSvcProxyName := rcsw.mirroredResourceName(spec.gatewayName)
+	service, err := rcsw.localAPIClient.Svc().Lister().Services(rcsw.serviceMirrorNamespace).Get(localSvcProxyName)
+	if err != nil {
+		return err
+	}
+
+	if service.Annotations != nil && service.Annotations[consts.RemoteGatewayResourceVersionAnnotation] != spec.resourceVersion {
+		updatedService := service.DeepCopy()
+		if updatedService.Annotations != nil {
+			updatedService.Annotations[consts.RemoteGatewayResourceVersionAnnotation] = spec.resourceVersion
+		}
+
+		endpoints, err := rcsw.localAPIClient.Endpoint().Lister().Endpoints(rcsw.serviceMirrorNamespace).Get(localSvcProxyName)
+		if err != nil {
+			return err
+		}
+
+		updatedEndpoints := endpoints.DeepCopy()
+		updatedEndpoints.Subsets = []corev1.EndpointSubset{
+			{
+				Addresses: spec.addresses,
+				Ports: []corev1.EndpointPort{
+					{
+						Protocol: "TCP",
+						Port:     int32(spec.ProbeConfig.port),
+					},
+				},
+			},
+		}
+
+		_, err = rcsw.localAPIClient.Client.CoreV1().Services(rcsw.serviceMirrorNamespace).Update(updatedService)
+		if err != nil {
+			return err
+		}
+
+		_, err = rcsw.localAPIClient.Client.CoreV1().Endpoints(rcsw.serviceMirrorNamespace).Update(updatedEndpoints)
+		if err != nil {
+			rcsw.localAPIClient.Client.CoreV1().Services(rcsw.serviceMirrorNamespace).Delete(updatedService.Name, &metav1.DeleteOptions{})
+			return err
+		}
+		rcsw.log.Debugf("%s gateway proxy service updated", localSvcProxyName)
+	}
+
+	return nil
+}
+
+func (rcsw *RemoteClusterServiceWatcher) createGatewayProxyService(spec *GatewaySpec) error {
+	localSvcProxyName := rcsw.mirroredResourceName(spec.gatewayName)
+	if spec.ProbeConfig == nil {
+		rcsw.log.Debugf("Skipping creation of gateway proxy service as gateway does not specify probe config")
+		return nil
+	}
+
+	_, err := rcsw.localAPIClient.Svc().Lister().Services(rcsw.serviceMirrorNamespace).Get(localSvcProxyName)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			serviceToCreate := &corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      localSvcProxyName,
+					Namespace: rcsw.serviceMirrorNamespace,
+					Labels: map[string]string{
+						consts.MirroredResourceLabel:  "true",
```

Comment:
I'm curious what we use this label for, if anything?  I ask because we now have two different kind of services that we create: mirrored services and gateway mirrors.  Anywhere that we're using this label selector will now select both kinds of services.  Is that appropriate?

Reply:
So, we use it in `cleanupMirroredResources` where we combine it with the cluster name. This deletes all mirrored services that needs to be deleted when we stop watching a cluster. We also use it in `mirroredServicesForGateway` but the re we combine it with some more specific annotations so there will be no conflict